### PR TITLE
acceptance-tests boiler plate

### LINKED ISSUE #28: create API tests
- [ ]  find appropriate API testing tools
- [ ] create example tests for ocis-hello
- [ ] run tests in drone

### PR DESCRIPTION
**Pull request by individual-it**
fixes #28 

After some research of other API testing tools and discussion with @phil-davis , I'm suggesting to stay with Gherkin style tests (against my previous assumption)
Gherkin seems to be 'best-practice' for this kind of API tests and more efficient that other tools.

For API tests I still suggest some changes in the Gherkin syntax to the usual UI tests:
1. don't bother to write full and nice English sentences, this are anyway technical terms
2. use tables (in the steps and in examples) as much a possible

This PR implements the tests in cucumber-js. The reasons to take the JS implementation and not the Go implementations are:

1. front-ends will be still written in JS, so dev. should be familiar with it
2. existing infrastructure can be shared with phoenix, where the tests are implemented in JS
3. not one more lib/toolstack in the "company" to know about

### CODE REVIEW

**Comment by tboerger** [comment]
Api tests are not acceptance tests, these are integration tests, that's how @micbar and me had chat about it.

Writing acceptance tests (the real user perspective on the ui) is fine to do in javascript because of Phoenix.

But I'm totally against writing integration tests with javascript.

1. It's much slower than golang 
2. Golang got a great set of libraries for testing 
3. Integration tests should be mostly written by the backend developers whose are writing golang and not javascript 
4. It's common to start the api handlers within the tests, so no need to start any service expect executing the test suite

**Comment by micbar** [comment]
@tboerger Good point.

Your suggesttion was to use goDog

@individual-it We were talking about goDog, does anything speak against ist?

I would also keep the integration tests close to the Developers.

**Comment by micbar** [comment]
@individual-it @tboerger We have a common understanding in the qa team (@HanaGemela @jnweiger)  that these API tests are Integration tests, independent from the fact that they could be written in Gherkin.

**Comment by tboerger** [comment]
A good reference for integration tests for go apis and handlers is this: https://github.com/drone/drone/blob/master/handler/api/badge/status_test.go

**Comment by individual-it** [comment]
@tboerger @micbar 
Can you explain what you understand under the term "integration-tests"? I thought of API tests as end-to-end black-box tests. In that case there are no mocks but the real (compiled & build) application started up and tested. That is how we currently test the API in oC10.

> We were talking about goDog, does anything speak against it?

1. sharing code
When the API gets more complex there will be common things to be done, that need to be done for UI (acceptance) tests and API (integration) tests. e.g. setting up users & groups, pre-create shares, setting specific settings in the system-under-test. Also checks would be partly same. e.g. check if a file was uploaded/modified/deleted, a calendar-item created, a user changed, etc. 
If we use JS, for this steps, we don't need to write them twice, but can share them between UI & API tests and even use existing steps from phoenix, that do exactly that.
Let's take a new extension that handles CalDav requests. We would still have to setup at least users and groups to test if sharing of calendars works. However that is done (ocs-api, LDAP manipulation, JSON file manipulation) it needs to be done for UI tests and for API tests and most-likely also for other apps. So why not share that code?

2. one more Lib/toolstack in the organisation to look after
The cucumber-js toolstack is already used and known in the company, one more in the list would just add complexity.

**Comment by tboerger** [comment]
I would personally even write the acceptance tests in go just because it's much faster than the whole javascript stuff.

And reuse stuff from Phoenix? Why should the Phoenix test code get reused for that?

**Comment by phil-davis** [comment]
My definition of (in an OO environment):
- unit test - tests the public methods of just a single class, independently (as much as possible) of the rest of the system
- integration test - tests the behaviour of multiple classes interacting with each other, but still with the classes getting "locally" instantiated
- acceptance test - the "whole system" is up and running with whatever infrastructure it would typically run in (e.g. a web server like Apache, a real backend database, real storage...) and the tests "hit the system" like a real client would. ("client" is anything that would use the external functionality of the system - somebody typing `curl` commands to the API, somebody using a webUI, a client program on a remote computer auto-syncing files...)

But these are just words and conventions. So we can define them how we want.

In this case we want API tests running against "the system in its production state".

**Comment by phil-davis** [comment]
If the alternate approaches are similar after assessing pros-cons, then the reason to use the nightwatch JS-based test framework is:
- we already use it on one project/repo and various developers and QA people are familiar with it
- choosing another test framework would mean developers and QA staff have to learn both frameworks (or we have to re-engineer the nightwatch JS already written for Phoenix across to the new framework

And we also have the Behat PHP-based framework that happened first, because core oC10 is a PHP-based backend, so that was chosen because developers writing PHP backend code will already be familiar with PHP.

The question somewhat, do we want a test framework for each product, using a language matching the language that the run-time of the product is written in = a few different frameworks, 1 per language; or do we want a standard blackbox-testing framework that we use for whatever system-under-test happens to come along?

File: tests/acceptance/features/hello.feature [hunk 1]
```diff
@@ -0,0 +1,48 @@
+Feature: hello API
+
+  As I user
```
Comment (phil-davis) [inline tests/acceptance/features/hello.feature]:
```suggestion
  As a user
```

File: tests/acceptance/features/hello.feature [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    Then response http status code should be 201
+    And response body should be
+      """
+      {"message":"Hello <name>"}
+
+      """
+    Examples:
+      | name         | comment                |
+      | Peter        | simple ASCII character |
+      | आर्तुर       | UTF caracters          |
```
Comment (phil-davis) [inline tests/acceptance/features/hello.feature]:
```suggestion
      | आर्तुर       | UTF characters          |
```

**Comment by tboerger** [comment]
> The question somewhat, do we want a test framework for each product, using a language matching the language that the run-time of the product is written in = a few different frameworks, 1 per language; or do we want a standard blackbox-testing framework that we use for whatever system-under-test happens to come along?

We should use a framework that best match the product/techstack, and for a product that is written in Go it's a Go testing framework for sure. You won't win other Go developers for implementing backend tasks if they got to write tests in Javascript...

Beside that, IMHO Gherkin is pretty much bloated for API testing. It's fine to define user stories, but API endpoints don't define user stories...

**Comment by individual-it** [comment]
I had a look at testing the API through the handler as proposed by @tboerger in https://github.com/owncloud/ocis-hello/pull/32#issuecomment-567477122
This seems not to be supported by swagger see https://goswagger.io/faq/faq_testing.html (but I confess the issue might also be my lack of understanding the system)

So I've made an alternative approach to do the tests in go see #37 
That approach still does end-to-end tests without any mocking but they are written in go, without any Gherkin. Adding Gherkin to it would be pretty straight forward, see https://github.com/DATA-DOG/godog/tree/master/examples/api

Looking at the discussion so far I can see multiple ways forward for this:

|no |how| advantages | disadvantages | example-implementation |
| -----| ---| --- | ---| --- |
|1 | Go through Handler | | - [only approach found so far is hacky](https://goswagger.io/faq/faq_testing.html#using-httptest) <br>- not end-to-end| NA |
|2 | Go e2e | - fast<br> - written in Go <br> - end-to-end (testing all layers, libs, etc)<br>- no extra layers for testing | - cannot reuse setup/check code from JS projects<br>- harder to read<br>- for UI tests a selenium lib need to be found | #37|
|3 | Go e2e with Cucumber | - easy to read<br> - fast, but probably slower than option 2.<br>- written in Go <br> - end-to-end | - one extra layer on top of 2.<br>- cannot reuse setup/check code from JS projects<br> - for UI tests a selenium lib need to be found<br>- one (two for UI projects) more lib. to learn | NA |
|3 | JS e2e with Cucumber | - easy to read<br>- end-to-end <br>- can reuse setup/check code from JS projects (less dev. time, less duplication)<br>- framework already known in the organisation |  - slow (how much difference does it make? it is only test-code, would not the API itself be the bottleneck?)<br>- extra layer (cucumber)<br>- extra language to learn for backend-only projects | #32 |

UI tests are mentioned just for the case that we want to have a similar approach for UI tests.
Please add/correct me if I have missed any option or any advantages/disadvantages

**Comment by tboerger** [comment]
My big question is, how much javascript code do we really need to share? The only pure javascript part would be Phoenix itself, but even here we will have ocis-phoenix with the go layer.

It could be an advantage to have the tests written in go as it performs much faster.

There are already multiple libraries which provide selenium bindings for performing the browser tests.

**Comment by individual-it** [comment]
Summary of the voice meeting with @micbar on 08.01.2020 for documentation:
for terminology clarification: 
- UI tests are end-to-end tests and acceptance-tests
- API tests are integration tests

general way forward:
- API tests don't need any Gherkin
- API tests will be implemented in Go
- continue working on #37 and try to use https://github.com/stretchr/testify / https://godoc.org/github.com/stretchr/testify
- ideally test the API through handlers, but if that is not possible we can compile and start the complete app as its done in #37 

- UI tests will have Gherkin
- UI tests will be implemented in JS
- UI for this app is not ready yet, so tests will be implemented when its availiable

**Comment by micbar** [comment]
@individual-it Thanks for the Summary.

**Pull request closed without merging**